flyteadmin/pkg/async/notifications/implementations/smtp_emailer.go

```diff
@@ -0,0 +1,138 @@
+package implementations
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/smtp"
+	"strings"
+
+	"github.com/flyteorg/flyte/flyteadmin/pkg/async/notifications/interfaces"
+	"github.com/flyteorg/flyte/flyteadmin/pkg/errors"
+	runtimeInterfaces "github.com/flyteorg/flyte/flyteadmin/pkg/runtime/interfaces"
+	"github.com/flyteorg/flyte/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyte/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/flyteorg/flyte/flytestdlib/logger"
+	"github.com/flyteorg/flyte/flytestdlib/promutils"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+)
+
+type SMTPEmailer struct {
+	config        *runtimeInterfaces.NotificationsEmailerConfig
+	systemMetrics emailMetrics
+	tlsConf       *tls.Config
+	auth          *smtp.Auth
+}
+
+func (s *SMTPEmailer) SendEmail(ctx context.Context, email admin.EmailMessage) error {
+
+	newClient, err := smtp.Dial(s.config.EmailerConfig.SMTPServer + ":" + s.config.EmailerConfig.SMTPPort)
```

Comment:
Just to confirm, this is the intended behavior?  If admin needs to send a thousand emails, it has to make a new client each time?

Reply:
Yes, this is intended behavior. When using the built-in email client of Go, most examples you find work that way. Also with the integration into the Notifications processor this is the simplest approach. Establishing a connection does not seem to be a very complex undertaking when looking at the source code of the email client. 

I agree that keeping the connection is more efficient. But it also makes the code more complex, because the connection state needs to be maintained over an instance - so whenever an email is meant to be sent, it first needs to be checked if the connection is still alive. 

If you want me to, I can refactor it to only open a connection once to the SMTP server.

Reply:
I just checked again the internal implementation of Golang's smtp client: It appears that the client implementation seems to be not multi-thread-safe. Every client instance keeps an internal state between function calls - so when reusing the same client while sending an email for writing another mail can lead to weird behavior. 

My assumption is that there is only a single email processor per Flyte Admin instance. That means that there should be no concurrent email processing on any Flyte Admin instance, which makes it safe to reuse an existing Flyste client.